### PR TITLE
Fix stale per-tile load indexes after coarse tiling propagation

### LINKED ISSUE #3003: Spyre coarse tiling mis-lowers unsqueeze consumer of amax result, causing incorrect denominator
### What happened?

A FlashAttention-like coarse-tiling case produces incorrect results when an `amax(..., dim=-1)` result is consumed through `unsqueeze(-1)` inside the same coarse-tiled fused kernel.

The issue appears to be in the generated layout for the `block_max.unsqueeze(-1)` input to a later `sub`. The `amax` output itself appears correct, but the subsequent read uses a device layout with an extra dimension that changes the physical row stride.

**Minimal Repro**
Script:

```text
test-spyre-scripts/repro_coarse_unsqueeze_denominator.py
```

Core compiled function:

```python
def denominator(q, k, mask):
    with spyre_hint(tiles={"B": 1}):
        with spyre_hint(tiles={"H": 1}):
            with spyre_hint(tiles={"Lq": 2}):
                with spyre_hint(tiles={"Lk": 1}):
                    with spyre_hint(work_div={"H": 4, "Lq": 8, "Lk": 8}):
                        scores = torch.matmul(q, k.transpose(-1, -2)) + mask
                        block_max = torch.amax(scores, dim=-1)
                        exp_scores = torch.exp(scores - block_max.unsqueeze(-1))
                        return exp_scores.sum(dim=-1)
```

The mask is all zeros, so it should not affect the math. It is included because it preserves the problematic fused/coarse-tiled graph shape.

**How To Run**
```bash
rm -rf /tmp/spyre_clean_repro_cache /tmp/spyre_clean_repro_debug

TORCHINDUCTOR_CACHE_DIR=/tmp/spyre_clean_repro_cache \
TORCH_COMPILE_DEBUG=1 \
TORCH_COMPILE_DEBUG_DIR=/tmp/spyre_clean_repro_debug \
SPYRE_INDUCTOR_LOG=1 \
SPYRE_INDUCTOR_LOG_LEVEL=debug \
python test-spyre-scripts/repro_coarse_unsqueeze_denominator.py
```

**Observed Result**
Example output:

```text
mismatches=3114/4096 max_abs=653.0
worst_index=(0, 3, 382) expected=289.75 actual=943.0
```

The generated logs show coarse tiling is active:

```text
coarse_tile_groups: 1 group(s) formed
LoopSpec(count=2)
```

**What To Look For**
Inspect the generated wrapper:

```bash
rg -n "device_size=\\[1, 512, 4, 2, 64\\]|LoopSpec|op='max'|op='sub'" \
  /tmp/spyre_clean_repro_cache -g '*.py'
```

Suspicious generated layout:

```python
# amax writes block_max with the expected tile layout:
device_size=[1, 512, 4, 64]
device_coordinates=[0, c1, c0, 0]

# later sub reads block_max.unsqueeze(-1) with an extra physical dimension:
device_size=[1, 512, 4, 2, 64]
device_coordinates=[0, c1, c0, 0, 0]
```

That extra `2` changes the physical stride of `c1`, so the `sub` appears to read the wrong row of `block_max`.

**Expected Result**
The compiled Spyre result should match the CPU reference within tolerance. The `unsqueeze(-1)` consumer of the `amax` result should preserve a layout compatible with the original `amax` output, rather than introducing a physical dimension that changes row addressing.

### What did you expect to happen?

the results should match

### How can we reproduce it?

Use this reproducer script
[repro_coarse_unsqueeze_denominator.py](https://github.com/user-attachments/files/29573009/repro_coarse_unsqueeze_denominator.py)

### Any environmental details we need to know?

No

### Anything else we need to know?

_No response_

### Relevant log output

```Shell
mismatches=3114/4096 max_abs=653.0
worst_index=(0, 3, 382) expected=289.75 actual=943.0
```

### PR DESCRIPTION
**Pull request by cyang49**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

This fixes a coarse tiling bug where a buffer’s layout could be retiled, but later loads from that buffer still used stale full-buffer stride coefficients.

In the flash attention repro, this affected sparse intermediates such as `amax(...).unsqueeze(-1)`. The generated wrapper could produce an incorrect extra tile dimension, for example:

```text
[1, 512, 4, 4, 64]
```

instead of the expected per-tile view:

```text
[1, 512, 4, 64]
```

**What Changed**
- Preserve retiled-buffer stride metadata returned by `_stamp_group()`.
- Delay retiled load-index patching until after `insert_tiling_propagation()`.
- Patch only exact same-loop consumers:
  - `loop_info.loop_group_id` must exactly match the stamped group id.
  - the op must actually read one of the retiled buffers.
- Rewrite stale full-stride load indexes to the corresponding per-tile strides, for example:

```text
2048*i1 + i2 -> 512*i1 + i2
```

- Refresh the replaced op body in:
  - the main `operations` list
  - `group_ops`
  - `V.graph.name_to_buffer`

**Why After Propagation**
`insert_tiling_propagation()` may insert copy/full-buffer ops and redirect outside consumers. Patching before propagation can rewrite the wrong consumer path. Patching after propagation lets the pass operate on the final op list and restrict the rewrite to exact same-loop consumers of retiled buffers.

#### Which issue(s) this PR is related to:

Fixes #3003

### CODE REVIEW

**Review by dgrove-oss — COMMENTED** [review]
PR #3066: Fix stale per-tile load indexes after coarse tiling propagation

Summary

After _stamp_group() retiles a buffer (shrinking its layout strides), consumers in the same loop group still used the original full-buffer stride coefficients in their load-index expressions. This PR captures the old/new strides from _divide_ranges(), delays the rewrite until after insert_tiling_propagation(), then walks exact same-loop consumers and rewrites the stale load indexes using a WrapperHandler.

Review

---
BLOCKERs

None.

---
SUGGESTIONs

- object.__setattr__ mutation before replace_computed_buffer_body (coarse_tile.py, _patch_retiled_load_indexes): The code does object.__setattr__(op.data, "inner_fn", new_inner_fn) and then passes op.data as new_data to replace_computed_buffer_body. This works today because op.data is the Pointwise/Reduction being handed in, but it's an unusual pattern — mutating the body's inner_fn in place and then handing it to the "replace" function. The existing pattern in insert_tiling_propagation (line ~1318) replaces inner_fn the same way, so it's consistent with the codebase. Fine as-is, but worth a one-line comment noting that the pre-mutation is intentional before the frozen-buffer replacement, for the same reason the comment at line 1025 of pass_utils.py explains the WrapperHandler contract.
- No regression test for the specific bug shape (tests/inductor/test_coarse_tile_e2e.py): The PR fixes a bug where amax(…).unsqueeze(-1) inside a tiled group produced a bogus extra tile dimension. The existing test_hint_flash_attention exercises max_running.unsqueeze(-2) which is the same pattern, but it was pre-existing and passes on main today — meaning either it was already broken in a way that didn't raise (wrong numerics masked by tolerances) or the exact shape/tiling combination that triggers the bug isn't exercised. A targeted unit test or a new e2e test using the minimal reproducer from the PR description (amax(dim=−1, keepdim=True) under coarse tiling, asserting the wrapper shape is [1, 512, 4, 64] not [1, 512, 4, 4, 64]) would close the loop. This is a meaningful gap for a correctness fix.
- stamped_group_id is recomputed in coarse_tile() instead of returned by _stamp_group() (coarse_tile.py:578): _stamp_group already computes nested_group_id = group_id + (0,) * (len(levels) - 1) internally and stamps it onto ops. The PR re-derives the same expression in coarse_tile() as stamped_group_id. If the formula in _stamp_group ever changes, the two can diverge silently. Consider having _stamp_group return (retiled_infos, nested_group_id) so the caller doesn't need to recompute it.
- _stride_rewrite_map uses Counter but Expr equality can surprise (coarse_tile.py:_stride_rewrite_map): The map deduplicates strides with old_counts[old] == 1 to avoid ambiguous rewrites. sympy.simplify is called before hashing, which is correct. However, sympy expressions aren't hashable with perfect algebraic equivalence (two syntactically different expressions that simplify to the same value can have different hashes). The unit tests cover the intended dedup behavior, so this is low risk in practice given the concrete-integer strides that _divide_ranges produces, but worth acknowledging.
- _retile_load_index_from_strides uses sorted(loop_vars, key=str) for determinism (coarse_tile.py:_retile_load_index_from_strides): Sorting by string representation of sympy symbols is reasonable but could theoretically produce unexpected orderings for symbols with numeric suffixes (e.g., c10 sorts before c9 lexicographically). If stride rewrites were ever partial (some vars not matching), the residual check would catch it. Fine for now.

---
QUESTIONs

- Interaction with _divide_reduction_ranges: _stamp_grion_ranges (line 1434) for reduction-level hints, but the PR only captures retiled info from _divide_ranges (non-reduction levels). Is it expected that reduction-dimension retiling never needs load-index patching, or is this a known gap?
                                                                                                                                          ---
Verdict                                                                                                                                   
APPROVE — the core fix is sound, the WrapperHandler pattern is the right approach and consistent with project conventions (no stale index expression rebuilding), and the unit tests for the newh. The three concerns above are minor to moderate polish items, none blocking.

**Comment by cyang49** [comment]
Thanks @dgrove-oss for the suggestions. I think we should address them in a follow-up PR and merge the bugfix asap.

**Comment by dgrove-oss** [comment]
We can't merge until CI recovers.  I think the e2e test that reproduces the error is worth adding to this PR since we have several hours (optimistically) until things are working again.

**Review by dgrove-oss — APPROVED** [review]
CI is functional again.

Approving, but generating an e2e reproducer test in a follow up is still worthwhile imo.
